File: php/packal.php
<?php
/**
 * This file contains a class to help you submit your workflows / themes (/ and reports) to
 * Packal.org. Use this code if you build your workflows with a custom build script and want
 * to add in automatic submission to Packal.org as part of the build. Check Packal.org for
 * documentation on how workflows need to be structured.
 *
 * See the `example.php` for usage examples of this class
 */


/**
 * Here are three URLs for Packal that are related to different servers.
 */

// Development
// define( 'PACKAL_BASE_API_URL', 'http://localhost:3000/api/v1/alfred2/' );

// Staging
define( 'PACKAL_BASE_API_URL', 'https://mellifluously.org/api/v1/alfred2/' );

// Production
// define( 'PACKAL_BASE_API_URL', 'https://www.packal.org/api/v1/alfred2/' );

class Packal {

	/**
	 * Constructs the submission POST request
	 *
	 * @param string $type   valid options: 'worfklow', 'theme', 'report'
	 * @param array $params  the submission parameters
	 * @param string $username your Packal.org username
	 * @param string $password your Packal.org password
	 */
	public function __construct( $type, $params, $username, $password ) {
		// These are the valid types to submit
		$types = [ 'workflow', 'theme', 'report' ];
		// Exit early if the type is not permitted
		if ( ! in_array( $type, $types ) ) {
			die( "$type is not a valid type. Valid types are: " . implode( ', ', $types ) );
		}
		$this->params = $params;
		// Standard setup
		$this->ch = curl_init();
		curl_setopt( $this->ch, CURLOPT_URL, PACKAL_BASE_API_URL . $type . '/submit' );
		curl_setopt( $this->ch, CURLOPT_RETURNTRANSFER, true );
		curl_setopt( $this->ch, CURLOPT_POST, true );
		curl_setopt( $this->ch, CURLOPT_SAFE_UPLOAD, true );

		// Call the submit method for the particular type
		if ( ! call_user_func_array( [ $this, $type ], [ $params ] ) ) {
			// Would should never get here because we'll exit earlier in the script
			die( 'Could not call user method' );
			// This should call an exception, maybe?
			return false;
		}

		// Add in username and password
		$this->postData = array_merge( $this->postData, [ 'username' => $username, 'password' => $password ] );
		$this->build_data();
	}

	/**
	 * Constructs a "Workflow" submission
	 *
	 * @param  array $params the submission parameters for the workflow
	 * @return bool
	 */
	public function workflow( $params ) {
		if ( ! $this->ensure_keys([ 'file', 'version' ], $params ) ) {
			return false;
		}
		$this->postData = [ 'workflow_revision' => $params ];
		return true;
	}

	/**
	 * Constructs a "Theme" submission
	 *
	 * @param  array $params the submission parameters for the report
	 * @return bool
	 */
	public function theme( $params ) {
		if ( ! $this->ensure_keys([ 'name', 'description', 'tags', 'uri' ], $params ) ) {
			return false;
		}
		$params['tags'] = implode( ',', $params['tags'] );
		$this->postData = [ 'theme' => $params ];
		return true;
	}

	/**
	 * Constructs a "Report" submission
	 *
	 * @param  array $params the submission parameters for the report
	 * @return bool
	 */
	public function report( $params ) {
		if ( ! $this->ensure_keys([ 'workflow_revision_id', 'report_type', 'message' ], $params ) ) {
			return false;
		}
		$this->postData = [ 'report' => $params ];
		return true;
	}

	/**
	 * Executes the submission
	 *
	 * @return string 	the results from the POST request
	 */
	public function execute() {
		// I should add some error handling in this
		$result = curl_exec( $this->ch );
		curl_close( $this->ch );
		return $result;
	}

	/**
	 * Builds the post fields
	 *
	 * The standard "http_build_query()" did not work for the nested file params, so we needed a
	 * different way to build the POST request for workflow submissions. See curl_custom_postfields
	 * below for the weird code there.
	 */
	private function build_data() {
		if ( isset( $this->postData['workflow_revision'] ) ) {
			$this->postData['workflow_revision[version]'] = $this->postData['workflow_revision']['version'];
			unset( $this->postData['workflow_revision'] );
			// Pass to a weird function that I did not write but that seems to work
			$this->curl_custom_postfields( $this->ch, $this->postData, [ $this->params['file'] ] );
		} else {
			// For themes and reports
			curl_setopt( $this->ch, CURLOPT_POSTFIELDS,  http_build_query( $this->postData ) );
		}
	}

	/**
	 * Ensures that relevant keys are available
	 *
	 * @param  array $keys   the expected parameters for sucessful submission
	 * @param  array $params the submission parameters passed
	 * @return bool
	 */
	private function ensure_keys( $keys, $params ) {
		foreach ( $keys as $key ) :
			if ( ! isset( $params[ $key ] ) ) {
				return false;
			}
		endforeach;
		return true;
	}

	// FROM PHP.NET, this is kind of a hacky way to get it to work, but it works!
	/**
	 * For safe multipart POST request for PHP5.3 ~ PHP 5.4.
	 *
	 * @param resource $ch cURL resource
	 * @param array $assoc "name => value"
	 * @param array $files "name => path"
	 * @return bool
	 */
	private function curl_custom_postfields($ch, array $assoc = array(), array $files = array()) {

	    // invalid characters for "name" and "filename"
	    static $disallow = array("\0", "\"", "\r", "\n");

	    // initialize body
	    $body = array();

	    // build normal parameters
	    foreach ($assoc as $k => $v) {
	        $k = str_replace($disallow, "_", $k);
	        $body[] = implode("\r\n", array(
	            "Content-Disposition: form-data; name=\"{$k}\"",
	            "",
	            filter_var($v),
	        ));
	    }

	    // build file parameters
	    foreach ($files as $k => $v) {
	        switch (true) {
	            case false === $v = realpath(filter_var($v)):
	            case !is_file($v):
	            case !is_readable($v):
	                continue; // or return false, throw new InvalidArgumentException
	        }
	        $data = file_get_contents($v);
	        $v = call_user_func("end", explode(DIRECTORY_SEPARATOR, $v));

	        // THIS IS A TERRIBLE HACK
	        $k = 'workflow_revision[file]';
	        list($k, $v) = str_replace($disallow, "_", array($k, $v));
	        $body[] = implode("\r\n", array(
	            "Content-Disposition: form-data; name=\"{$k}\"; filename=\"{$v}\"",
	            "Content-Type: application/octet-stream",
	            "",
	            $data,
	        ));
	    }

	    // generate safe boundary
	    do {
	        $boundary = "---------------------" . md5(mt_rand() . microtime());
	    } while (preg_grep("/{$boundary}/", $body));

	    // add boundary for each parameters
	    array_walk($body, function (&$part) use ($boundary) {
	        $part = "--{$boundary}\r\n{$part}";
	    });

	    // add final boundary
	    $body[] = "--{$boundary}--";
	    $body[] = "";

	    // set options
	    return curl_setopt_array($ch, array(
	        CURLOPT_POST       => true,
	        CURLOPT_POSTFIELDS => implode("\r\n", $body),
	        CURLOPT_HTTPHEADER => array(
	            "Expect: 100-continue",
	            "Content-Type: multipart/form-data; boundary={$boundary}", // change Content-Type
	        ),
	    ));

	}
} // End Submit class

File: ruby/packal.rb
require 'rest-client'
require 'json'


# The Packal class to submit to the Packal API.
#
# Example Use:
#
# Submit a Workflow:
# =================
# json = {
# 	:workflow => {:file => '/path/to/workflow/file/workflow.alfredworkflow', :version => '1.0.0'},
# 	:username => "USERNAME",
# 	:password => "PASSWORD"
# }
# output = Packal.queue(json.to_json)
#
# Submit a Theme:
# ==============
# json = {
# 	:theme => {
# 		:name => 'Testing, one, two, three',
# 		:description => 'This is a description',
# 		:uri => 'alfred://theme/....',
# 		:tags => ['tag_one', 'tag_2']
# 		},
# 	:username => "USERNAME",
# 	:password => "PASSWORD"
# }
# output = Packal.queue(json.to_json)
#
# Submit a Report:
# ===============
# json = {
# 	:report => {
# 		:type => 'Malicious Code',
# 		:workflow_revision_id => 534,
# 		:message => 'There is some bad code in this thing, yo.'
# 		},
# 	:username => "USERNAME",
# 	:password => "PASSWORD"
# }
# output = Packal.queue(json.to_json)
#
class Packal
	def self.queue(json)
		begin
			# Start everythng up and parse the JSON first
			Packal.send(JSON.parse(json))
		rescue StandardError => e
			# If there is an error, then print the error
			e.message
		end
	end

	# Actually submit the data to Packal
	def self.submit(params)
		type = ('workflow_revision' == params.keys[0].to_s) ? 'workflow'.to_sym : params.keys[0]

		# Development Server
		# RestClient::Request.execute(:url => "http://localhost:3000/api/v1/alfred2/#{type}/submit", :payload => params, :method => :post) # development

		# Staging Server
		RestClient::Request.execute(:url => "https://mellifluously.org/api/v1/alfred2/#{type}/submit", :payload => params, :method => :post)

		# Production Server
		# RestClient::Request.execute(:url => "https://www.packal.org/api/v1/alfred2/#{type}/submit", :payload => params, :method => :post)
	end

	def self.send(params)
		# This is the type that is being sent through the metaclass
		type = params.keys[0]
		# Check to make sure that we are sending a USERNAME and a PASSWORD
		['username', 'password'].each do |k|
			unless (params.type? k) then
				raise StandardError.new("Error: you need to pass a #{k} to submit a #{type}.")
			end
		end

		# So, we'll make sure we use the correct submission process by using the submit method
		# of the appropriate type class. See below.
		Packal::const_get(type.capitalize).submit(params)
	end

	# Class to submit a report to Packal.org
	class Report
		def self.submit(params)
			# Make sure that we have the keys necessary for the report
			Packal.ensure_keys(params, ['workflow_revision_id', 'report_type', 'message'])
			# Submit through the Packal class
			Packal.submit(params)
		end
	end

	# Class to submit a theme to Packal.org
	class Theme
		def self.submit(params)
			# Make sure that we have all the keys necessary for the theme
			Packal.ensure_keys(params, ['name', 'description', 'uri'])
			# Submit through the Packal class
			Packal.submit(params)
		end
	end

	# Class to submit a Workflow to Packal.org
	class Workflow
		def self.submit(params)
			# Make sure that we have the "file" and "version" keys set
			Packal.ensure_keys(params, ['file', 'version'])
			# Make sure that the file exists
			file = ensure_file(params)
			# Make sure that the file has a zip mime-type
			check_mime_type(file)
			# Open the file as a resource so that we send the data
			data = File.new(file, 'rb')
			# Submit the workflow
			Packal.submit(fix_keys(params, data))
		end

		# Private Methods for the Workflow Class
		private

			# Ensures that the file to upload exists
			def self.ensure_file(params)
				file = params['workflow']['file']
				if ! File.exist?(file) then
					raise StandardError.new("Error: file #{file} not found.")
				end
				file
			end

			# Checks to make sure that the mimetype of the file is a valid archive
			def self.check_mime_type(file)
				unless ('zip' == `file --mime -b "#{file}"`.split("/")[1].split(';')[0])
					raise StandardError.new('Error: workflow file is not a valid archive.')
				end
			end

			# Fixes the keys so that they work correctly
			def self.fix_keys(params, data)
				# The problem is that the person using this library wants to submit a "workflow" but the
				# data on Packal.org is actually a workflow_revision, and so for the params to work correctly
				# we will just have to reshuffle the parameters.

				# Add the workflow_revision parameter
				params['workflow_revision'] = { :file => data, :version => params['workflow']['version'] }
				# Delete the workflow parameter
				params.delete('workflow')
				# Shuffle the keys a bit to make them play nice with the rest of the class because 'workflow_revision'
				# needs to be the first index
				keys = params.keys.reverse.map(&:to_sym)
				params = params.to_a.reverse.map{|x| x[1]}
				# Send the fixed parameters
				Hash[keys.zip(params)]
			end
	end

	# Private methods for the Packal class
	private
		# This method makes sure that the keys are present in the params Hash
		def self.ensure_keys(params, keys)
			# The first key at index '0' will be the 'type'
			type = params.keys[0]
			keys.each do |k|
				unless (params[type].key? k) then
					raise StandardError.new("Error: you need to pass a #{k} to submit a #{type}.")
				end
			end
			true
		end
end